### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/TensorProduct/Finiteness): add some finiteness results of tensor product

### PR DESCRIPTION
**Pull request by acmepjz**
- `TensorProduct.exists_multiset`, `TensorProduct.exists_finsupp_left`,
  `TensorProduct.exists_finsupp_right`, `TensorProduct.exists_finset`:
  any element of `M ⊗[R] N` can be written as a finite sum of pure tensors.
  See also `TensorProduct.span_tmul_eq_top`.

- `TensorProduct.exists_finite_submodule_left_of_finite`,
  `TensorProduct.exists_finite_submodule_right_of_finite`,
  `TensorProduct.exists_finite_submodule_of_finite` and 3 more:
  any finite subset of `M ⊗[R] N` is contained in `M' ⊗[R] N'`
  for some finitely generated submodules `M'` and `N'` of `M` and `N`, respectively.
  Each of these 3 functions has 2 variants.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+  for some finitely generated submodules `M'` and `N'` of `M` and `N`, respectively.
+  Each of these 3 functions has 8 variants.
+
+## Tags
+
+tensor product, finitely generated
+
+-/
+
+open scoped Classical TensorProduct
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean:33]:
```suggestion
open scoped TensorProduct
```
This breaks a couple of proofs, but it is enough to start them with `classical`.

File: Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+
+open Submodule
+
+noncomputable section
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean:37]:
```suggestion
```
This seems useless (without the `Classical` above).

File: Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+theorem exists_finset (x : M ⊗[R] N) :
+    ∃ S : Finset (M × N), x = S.sum fun i ↦ i.1 ⊗ₜ[R] i.2 := by
+  obtain ⟨S, h⟩ := exists_finsupp_left x
+  use S.graph
+  rw [h, Finsupp.sum]
+  refine' Finset.sum_nbij' (fun m ↦ ⟨m, S m⟩) Prod.fst .. <;> simp
+
+/-- If `x : ι → M ⊗[R] N` are finitely many elements, then there exists a finitely generated
+submodule `M'` of `M` and `x' : ι → M' ⊗[R] N`, whose image in `M ⊗[R] N` is equal to `x`.
+There are 8 variants of this function. -/
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean:94]: Do we really want 8 versions of this result? I understand that finding a single canonical one is probably not possible, but 8 seems really a lot.

Comment (erdOne) [reply Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean:94]: I think 
```
example (s : Set (M ⊗[R] N)) (hs : s.Finite) : 
    ∃ (M' : Submodule R M) (N' : Submodule R N), M'.FG ∧ N'.FG ∧
      s ⊆ (TensorProduct.map M'.subtype N'.subtype).range
```
is enough?

Comment (erdOne) [reply Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean:94]: Also I don't think the lemmas above are necessary. Usually when you find yourself writing tensors as a sum of pure tensors in lean, it is usually better done directly using induction.
```lean
import Mathlib

variable {R M N : Type*}

variable [CommSemiring R] [AddCommMonoid M] [AddCommMonoid N] [Module R M] [Module R N]

variable {M₁ : Submodule R M} {N₁ : Submodule R N}

open TensorProduct Submodule

lemma TensorProduct.range_mapIncl (M' : Submodule R M) (N' : Submodule R N) :
    LinearMap.range (mapIncl M' N') = span R (Set.image2 (· ⊗ₜ ·) M' N') := by
  rw [mapIncl, map_range_eq_span_tmul]
  congr; ext; simp

lemma TensorProduct.range_mapIncl_mono
    {M₁' M₂' : Submodule R M} {N₁' N₂' : Submodule R N}
    (hM : M₁' ≤ M₂') (hN : N₁' ≤ N₂') :
    LinearMap.range (mapIncl M₁' N₁') ≤ LinearMap.range (mapIncl M₂' N₂') := by
  rw [range_mapIncl, range_mapIncl]
  exact span_mono (Set.image2_subset (SetLike.le_def.mp hM) (SetLike.le_def.mp hN))

example (s : Set (M ⊗[R] N)) (hs : s.Finite) : 
    ∃ (M' : Submodule R M) (N' : Submodule R N), M'.FG ∧ N'.FG ∧
      s ⊆ LinearMap.range (mapIncl M' N') := by
  refine hs.induction_on ?_ ?_
  · exact ⟨_, _, fg_bot, fg_bot, Set.empty_subset _⟩
  rintro a s - - ⟨M', N', hM', hN', h⟩
  induction' a using TensorProduct.induction_on with x y x y hx hy
  · exact ⟨M', N', hM', hN', Set.insert_subset (zero_mem _) h⟩
  · refine ⟨_, _, hM'.sup (fg_span_singleton x),
      hN'.sup (fg_span_singleton y), Set.insert_subset ?_ ?_⟩
    · exact ⟨⟨x, mem_sup_right (mem_span_singleton_self x)⟩ ⊗ₜ
        ⟨y, mem_sup_right (mem_span_singleton_self y)⟩, rfl⟩
    · exact fun z hz ↦ range_mapIncl_mono le_sup_left le_sup_left (h hz)
  · obtain ⟨M₁', N₁', hM₁', hN₁', h₁⟩ := hx
    obtain ⟨M₂', N₂', hM₂', hN₂', h₂⟩ := hy
    refine ⟨_, _, hM₁'.sup hM₂', hN₁'.sup hN₂', Set.insert_subset (add_mem ?_ ?_) ?_⟩
    · exact range_mapIncl_mono le_sup_left le_sup_left (h₁ (Set.mem_insert x s))
    · exact range_mapIncl_mono le_sup_right le_sup_right (h₂ (Set.mem_insert y s))
    · exact fun z hz ↦ range_mapIncl_mono le_sup_left le_sup_left
        (h₁ (Set.subset_insert x s hz))
```

Comment (acmepjz) [reply Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean:92]: > Do we really want 8 versions of this result? I understand that finding a single canonical one is probably not possible, but 8 seems really a lot.

Let me see. Currently there are the following combinations:

- submodules: left, right or both
- M and N are types with module structure, or they are already submodules
- fintype, finsupp, pair, or only one element. From the proof I think the latter two cases can be removed. Also the finsupp case is a possible variation, but it's never used yet.

In fact what I really need is `exists_finite_submodule_left_of_pair'` and `exists_finite_submodule_right_of_pair'`, used in https://github.com/leanprover-community/mathlib4/blob/2d07f5a79f80dd8d4371b2bc02267ec03b3c88c5/Mathlib/LinearAlgebra/LinearDisjoint.lean#L483-L502

Comment (acmepjz) [reply Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean]: OK, now there are only 6 results of `exists_finite_submodule`. Adapted the proof of @erdOne. Please give some comments.

> And I don't think there is need for left and right versions.

Unfortunately I need to use them in linear disjoint file... However now left and right versions are corollaries of the both version.

Also the `exists_multiset`, etc. are not removed yet. I think they *may be* useful by someone...

File: Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    ∃ S : Finset (M × N), x = S.sum fun i ↦ i.1 ⊗ₜ[R] i.2 := by
+  obtain ⟨S, h⟩ := exists_finsupp_left x
+  use S.graph
+  rw [h, Finsupp.sum]
+  refine' Finset.sum_nbij' (fun m ↦ ⟨m, S m⟩) Prod.fst .. <;> simp
+
+lemma range_mapIncl_mono (hM : M₁ ≤ M₂) (hN : N₁ ≤ N₂) :
+    LinearMap.range (mapIncl M₁ N₁) ≤ LinearMap.range (mapIncl M₂ N₂) := by
+  simp_rw [mapIncl, ← subtype_comp_inclusion _ _ hM, ← subtype_comp_inclusion _ _ hN, map_comp]
+  exact LinearMap.range_comp_le_range _ _
```
Comment (erdOne) [inline Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean:98]: This should go under the definition of `mapIncl`.

File: Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean (hunk 3 above)
Comment (erdOne) [reply Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean]: I originally thought that the `LinearMap.range_comp_le_range` part could be performed inline at the application site and there is no need for separate lemmas concerning left and right. But as long as the number of lemmas does not overwhelm the user, we might as well provide them for convenience.

**Comment by erdOne** [comment]
LGTM. I'll wait if Riccardo wants to give it a look, and I'll maintainer-merge this tomorrow otherwise.

**Comment by riccardobrasca** [comment]
Sorry, I was a little busy. LGTM too, thanks!

bors merge

**Pull request closed without merging**